### ISSUE TITLE
MacOS Media Player integration (media keys/playback status)

### ISSUE THREAD

**Issue by subpanic** [jeffvli/feishin#759]
### What do you want to be added?

Currently Feishin's playback status and controls aren't integrated with the Media Player API in MacOS. This means on devices like Macbooks the media keys cannot currently control Feishin and Feishin's playback state and status (artist/song name) don't show up in the MacOS Control Center. (Note: this is different than desktop MacOS setups with some external keyboards, playback controls can still work in some instances there when the keys aren't tied to the Media Player controls, but they are on Mac laptops)

I much prefer Feishin's app but have been forced to use alternatives on my macbook because of lack of media key control.

### Is this a server-specific feature? (e.g. Jellyfin only)

- [ ] Yes

**Comment by jamie6king** [comment]
I'm fairly certain you need to enable Feishin in Settings > Privacy & Security > Accessibility. I have a MacBook and it works perfectly (although it doesn't appear in the menubar for me).

**Comment by subpanic** [comment]
This is true but not the issue. This could be related to MacOS version/macbook version, I'm on 14.5 on an M1 Macbook Pro. Media keys will only trigger iTunes unless another application is running integrated with the Media Player API, it's a fairly widely documented issue. As a sidenote Feishin does have accessibility access, but the media keys will get captured by whatever application is responding to the Media Player API (which will be iTunes my default if no other running app is registered).

This was also previously an issue with the Supersonic application on certain MacOS/Mac hardware version until they added Media Player API integration.

**Comment by subpanic** [comment]
Follow-up, I did some further testing (removing the app, unsetting and re-setting accessibility, etc.) and was able to get media keys working. For anyone that has a similar issue where media keys are stolen by iTunes/Music app:

- Toggle accessibility for Feishin off and then on again
- Following this, In the Feishin app go to Settings > Hotkeys and toggle "Global media hotkeys" off and then on again. This setting must be re-toggled after enabling accessibility

Leaving this open as Media Player API would still improve experience and remove/simplify this issue for people. But a maintainer can feel free to close/lmk to close if they would prefer a fresh feature ticket.

**Comment by ceilingtilefan** [comment]
I can't seem to get media keys working at all and I've _never_ been able to get them working - iTunes always takes priority for some reason. Quite annoying when using AirPods while away from my computer

**Comment by subpanic** [comment]
@bigfloppa234 As a possible interim solution for you (not ideal but it's something) you can try noTunes (https://github.com/tombonez/noTunes)

Basically run noTunes and it'll have a little icon in your top taskbar, it prevents the Music app from ever starting/stealing keys. Then follow the steps above for re-toggling accessibility and global media hotkeys. I believe that should get it working for you (including buttons on headphones).

**Comment by ceilingtilefan** [comment]
> @bigfloppa234 As a possible interim solution for you (not ideal but it's something) you can try noTunes (https://github.com/tombonez/noTunes)
> 
> 
> 
> Basically run noTunes and it'll have a little icon in your top taskbar, it prevents the Music app from ever starting/stealing keys. Then follow the steps above for re-toggling accessibility and global media hotkeys. I believe that should get it working for you (including buttons on headphones).

Odd, this doesn't work for me either. noTunes kills Music (verified by pressing play on my keyboard) but no media shortcuts work on Feishin. 

Edit: Using macOS 15.1 (24B5035e)

**Comment by anthonymag** [comment]
Still no dice for me either unfortunately. On top of the media keys not working, the media icon in the upper taskbar doesn't show Feishin activity, and I'm fairly certain they go hand in hand with the music API.

I've seen others report they got the keys working, but that the taskbar icon still doesn't show activity, which makes me think the hotkeys solution is likely a hack/override rather than a proper use of the API.

**Comment by TalhaMangarah** [comment]
@ceilingtilefan @anthonymag 

To get media keys to work, I had to remove Feishin from the accessibility list by selecting Feishin then clicking the '-'. To add it back, click on the '+' and add the program from the applications list. Just toggling off and on did not work for me from the accessiblity list in Privacy & Security but now it's working.

**Comment by anthonymag** [comment]
@TalhaMangarah you're a legend, that actually worked for me, thanks!

As I suspected above though, playback continues not showing in the media panel, so I'm pretty sure it isn't using the API properly and the key functions are an override.

**Comment by subpanic** [comment]
Yeah it won't show playback in the media panel. This isn't a bug, it's just not implemented. That's what this enhancement ticket was originally created for, properly implementing support for the Media Player API. Doing so would make it so that Feishin works like most other media players do on MacOS (integrated system-wide playback status/media functions).

**Comment by manicmoose** [comment]
Weird....On MacOS 0.12.3 I can get fwd & back keys to work, but play/pause will not.🤨
Have tried all the suggestions above.

**Comment by Xiione** [comment]
I've successfully forked and built the app locally with macOS MediaRemote support. That is, I'm able to control the app with the system media keys (with the in-app "Global media hotkeys" setting disabled), and the current track's art, title, album, artist(s), etc. will show in the menu bar widget. 
If anyone is still interested in this feature let me know here and I'll set up Github Actions on my fork repo

**Comment by manicmoose** [comment]
@Xiione Nice work, but couldn't you just submit a PR...if it's not that hard, I can't imagine it wouldn't get merged.
How much code change was required?

**Comment by Xiione** [comment]
@manicmoose I'll definitely submit a PR, just not sure if this repo is still getting looked at since it's being rewritten as audioling. Another thing is that I had to update and install some packages to get it to build, I don't want to contaminate upstream if it's a me issue. As for the code, I used [electron-media-service](https://github.com/Xiione/electron-media-service) to implement as much of the logic already existing for linux as possible. The changes are isolated to the macos build. ([the commit](https://github.com/Xiione/feishin/commit/fcf1a947e27558b4f3ac26b7830f0370d0b85ffd))

**Comment by subpanic** [comment]
@Xiione a release for your fork would be incredible since we can't be sure on state of getting a new PR in this repo. Thanks for your work on this! (or alternatively if you could just outline build steps that would also be great)

**Comment by Xiione** [comment]
@manicmoose @subpanic In trying to make the fork PR-ready I unfortunately kept running into weird CI build issues triggered by adding a new package. I don't currently have time to find a way to fix them, but I do have a specific build procedure that should work to build it locally on your Mac machine.

Using Node v18.20.8 (I used fnm to change node versions):
1. Clone [the fork](https://github.com/Xiione/feishin)
2. In the root directory (`feishin/`), run `npm i --legacy-peer-deps`
3. After installation, edit `release/app/package.json` such that `electron` is in `devDependencies` instead of `dependencies`:

Before:
```json
    "dependencies": {
        "electron": "^33.3.1",
        "@xiione/electron-media-service": "^0.2.3",
        "cheerio": "^1.0.0-rc.12",
        "mpris-service": "^2.1.2",
        "ws": "^8.18.0"
    },
    "devDependencies": {
    },

```
After:
```json
    "dependencies": {
        "@xiione/electron-media-service": "^0.2.3",
        "cheerio": "^1.0.0-rc.12",
        "mpris-service": "^2.1.2",
        "ws": "^8.18.0"
    },
    "devDependencies": {
        "electron": "^33.3.1"
    },
```
4. Finally, run `npm run package` in the root directory.
5. Under `release/build/mac` and `release/build/mac-arm64` should be the built app bundles. Note that I only own an arm64 mac so I don't know if things are working properly on Intel macs.
6. In Feishin, you can now **disable** `Settings > Hotkeys > Global media hotkeys`. When you play other media after pausing feishin, using the media keys should respect the new "current" media.

The fix is definitely a weird one but after trying for a bit to get a CI-built release I could see why the author is looking to rewrite the project. Absolutely no disrespect though, no other music player app built for macOS seems to meet my needs as perfectly as Feishin does.

Let me know if any part of the procedure doesn't work on your machine and I'll do my best to help and update the steps.

**Comment by subpanic** [comment]
Can confirm working! Awesome work @Xiione 

Only additional note is for anyone using later node/npm versions, don't try. You need to make sure you are using the correct node version. I had to clear my cache `npm cache clean --force` after installing v18.20.8 with fnm (unsure if this is me doing stupid things and fnm has a better way to deal with npm cache between versions, node/npm isn't my usual space).

**Comment by GlassedSilver** [comment]
The lack of proper integration of the API macOS provides comes with another caveat. Apps that look for media playback won't recognize Feishin.

This is an issue for apps like https://github.com/ungive/discord-music-presence which I would love to use in conjunction with Feishin, since it arguably is the best way to have a consistent experience for scrobbling to Last.fm and Discord Rich Presence.

**Comment by 4rft5** [comment]
> The lack of proper integration of the API macOS provides comes with another caveat. Apps that look for media playback won't recognize Feishin.


Agreed, though I'd imagine it'd be argued to just use the Feishin Discord presence, the point still stands.

I use an external tool called MediaMate to display MacOS Now Playing which comes with the advantage of being able to control playback and songs from the notch, and that doesn't work either because there's no support.

I tried the build in here that claims to add this, but it didn't work at all for me. Maybe that's just for the media controls and not actual Now Playing API, but it's still something that needs to be added. I hope this issue is reopened (shouldn't have been closed to begin with) so it can be added.

**Comment by jeffvli** [comment]
This should be solved though.
If you enable mediasession in the hotkey settings, does it not work?

**Comment by 4rft5** [comment]
> This should be solved though. If you enable mediasession in the hotkey settings, does it not work?

I hadn't tried the Media Session setting before because in its description it says Windows only, but it does seem to be working as Now Playing with MediaMate/MacOS natively, which is awesome. Thank you!

@GlassedSilver Maybe this will work for you too?

**Comment by GlassedSilver** [comment]
Holy heck, the media session works on macOS as well. I hope the info text is updated to reflect it's Windows & macOS.

Now all that's left is for music-presence to add Feishin's macOS builds to the detected player list. :)

**Comment by GlassedSilver** [comment]
Update: it all works beautifully now!

Music Presence is able to one-stop-shop handle all my Last.fm scrobbles and my Discord Rich Presence for YouTube Music, Apple Music and my beloved Feishin!

The macOS build just received an update to enable toggling on Feishin on Mac recognition. (for anyone affected: just restart Music Presence, definitions seem to be pulled on every application launch! :) )

This is an absolute win!

**Comment by manicmoose** [comment]
Enabling "media session" works for fwd/back, but not for play/pause - even with a cabled Apple keyboard.
Anything else I can try, @jeffvli ??

**Comment by jeffvli** [comment]
@manicmoose I can't answer that, as I do not have a macOS device to test any of this functionality.

MediaSession is only strictly necessary if you want the integration with your OS media controls. If you just want global hotkeys, you can just use the `Enable global hotkeys` toggle, and then make to set Feishin as a Trusted Accessibility Client.

**Comment by 4rft5** [comment]
> Enabling "media session" works for fwd/back, but not for play/pause - even with a cabled Apple keyboard.

It works for me just fine. Make sure it's enabled in accessibility settings (you may have to remove with the minus button and re-add).

**Comment by nathan-contino** [comment]
For anyone else who lost media keys access in 1.3 or 1.4, I can confirm that "global media keys" works on macOS if you remove the permission with the minus button and then manually re-add Feishin in Privacy & Security > Accessibility. Just toggling it off and on did not work for me, I had to actually remove it.
